torch/nn/utils/_stateless.py

```diff
@@ -1,43 +1,130 @@
+    # Changes the module class to get a new __getattr__ dunder method
+    # that looks for the reparametrized tensor
+    if hasattr(module, "_functional_parameters"):
+        module._functional_parameters[tensor_name] = tensor
+    else:
+        module._functional_parameters = {}
+        module._functional_parameters[tensor_name] = tensor
+        _change_class(module, apply_parametrizations)
+
+
+def _remove_swap(module, name: str) -> None:
+    if hasattr(module, "_orig_class"):
+        module.__class__ = module._orig_class
+        delattr(module, "_orig_class")
+        delattr(module, "_functional_parameters")
 
 
 @contextlib.contextmanager
-def reparametrize_module(module, parameters_and_buffers):
-    # Parametrization does not support to change submodules directly
+def reparametrize_module(
+    module: torch.nn.Module,
+    parameters_and_buffers: Dict[str, Tensor],
+    apply_parametrizations: bool
+) -> Iterator[None]:
     for name, tensor in parameters_and_buffers.items():
         _apply_func_submodules(
-            torch.nn.utils.parametrize.register_parametrization,
-            module, name.split("."), (_ReparametrizedTensor(tensor),))
+            _swap_parameters,
+            module, name.split("."), (tensor, apply_parametrizations))
     yield
     for name in parameters_and_buffers:
         _apply_func_submodules(
-            torch.nn.utils.parametrize.remove_parametrizations,
-            module, name.split("."), (False,))
-
+            _remove_swap,
+            module, name.split("."), ())
 
-class _ReparametrizedTensor(torch.nn.Module):
-    def __init__(self, tensor):
-        super().__init__()
-        self._tensor = tensor
 
-    def forward(self, original):
-        return self._tensor
-
-
-def _apply_func_submodules(func, module, path, args):
+def _apply_func_submodules(
+    func: Callable[..., None],
+    module: torch.nn.Module,
+    path: List[str],
+    args: Tuple,
+):
     if len(path) == 1:
         func(module, path[0], *args)
     else:
         _apply_func_submodules(func, getattr(module, path[0]), path[1:], args)
 
 
-def functional_call(module, parameters_and_buffers, args, kwargs=None):
+def functional_call(
+    module: torch.nn.Module,
+    parameters_and_buffers: Dict[str, Tensor],
+    args: Tuple,
+    kwargs : Dict[str, Any] = None,
+    *, apply_parametrizations : bool = False,
+):
+    r"""Performs a functional call on the module by replacing the module parameters with
+    the provideed ones.
+
+    Args:
+        module (torch.nn.Module): the module to call
+        parameters_and_buffers (dict of str and Tensor): the parameters that will be used in
+            the module call.
+        args (tuple): arguments to be passed to the module call
+        kwargs (dict): keyword arguments to be passed to the module call
+        apply_parametrizations (bool): if ``True`` and ``module`` has parametrizations registered
```

Comment:
Do we actually want this flag?
I feel like the user already controls that by either overriding `mod.weight` or `mod.parametrizations.weight.original` ?

The other thing is that it is easier to add this flag afterwards if the above assumption doesn't hold in practice than removing it if it ends up being useless.

Reply:
I originally thought of it as an easy way for users to avoid having to set the key to `mod.parametrizations.weight.original` when they want to override parametrization underlying tensor.
But yeah, is redundant so let's get rid of it now and add it later if we feel it is needed

Reply:
Pushed!